nixos/tests/pangolin.nix

```
@@ -0,0 +1,134 @@
+  # setup ACME server to sign certificates and point
+  # DNS to the correct machine
+
+  nodes = {
+    # The fake ACME server which will respond to client requests
+    acme =
+      { nodes, ... }:
+      {
+        imports = [ ./common/acme/server ];
+        networking.nameservers = lib.mkForce [ (dnsServerIP nodes) ];
+      };
+
+    # A fake DNS server which can be configured with records as desired
+    # Used to test DNS-01 challenge
+    dnsserver =
+      { nodes, ... }:
+      {
+        networking = {
+          firewall.allowedTCPPorts = [
+            8055
+            53
+          ];
+          firewall.allowedUDPPorts = [ 53 ];
+
+          # nixos/lib/testing/network.nix will provide name resolution via /etc/hosts
+          # for all nodes based on their host names and domain
+          hostName = "dnsserver";
+          domain = "eu";
+        };
+        systemd.services.pebble-challtestsrv = {
+          description = "Pebble ACME challenge test server";
+          wantedBy = [ "network.target" ];
+          serviceConfig = {
+            ExecStart = "${lib.getExe' pkgs.pebble "pebble-challtestsrv"} -dns01 ':53' -defaultIPv6 '' -defaultIPv4 '${nodes.VPS.networking.primaryIPAddress}'";
+            # Required to bind on privileged ports.
+            AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+          };
+        };
+      };
+
+    VPS =
+      { nodes, ... }:
+      {
+        imports = [ ./common/acme/client ];
+        networking = {
+          inherit domain;
+          hosts.${nodes.VPS.networking.primaryIPAddress} = [
+            domain
+            "pangolin.${domain}"
+          ];
+          nameservers = lib.mkForce [ (dnsServerIP nodes) ];
+        };
+
+        environment = {
+          etc = {
+            "nixos/secrets/pangolin.env".text = ''
+              SERVER_SECRET=${secret}
+            '';
+          };
+        };
+
+        services = {
+          pangolin = {
+            enable = true;
+            baseDomain = domain;
+            letsEncryptEmail = "pangolin@${domain}";
+            openFirewall = true;
+            environmentFile = "/etc/nixos/secrets/pangolin.env";
+            settings = {
+              flags.enable_integration_api = true;
+            };
+          };
+          # set up local ca server, so we can get our certs signed without going on the internet
+          traefik.staticConfigOptions.certificatesResolvers.letsencrypt.acme.caServer =
+            lib.mkForce "https://${nodes.acme.test-support.acme.caDomain}/dir";
+        };
+      };
+
+  };
+  testScript = ''
+    start_all()
+
+    with subtest("start pangolin}"):
+      VPS.wait_for_unit("pangolin.service")
+      VPS.wait_for_open_port("3002")
+
+    with subtest("start gerbil}"):
+      VPS.wait_for_unit("gerbil.service")
+      VPS.wait_for_open_port("3003")
+
+    with subtest("start traefik}"):
+      VPS.wait_for_unit("traefik.service")
+
+    with subtest("check traefik certs}"):
```

```suggestion
    with subtest("start pangolin"):
      VPS.wait_for_unit("pangolin.service")
      VPS.wait_for_open_port("3002")

    with subtest("start gerbil"):
      VPS.wait_for_unit("gerbil.service")
      VPS.wait_for_open_port("3003")

    with subtest("start traefik"):
      VPS.wait_for_unit("traefik.service")

    with subtest("check traefik certs"):
```

Should we also check if ports 80 and 443 are opened?

---

Sure, why not.